### PR TITLE
Add a control pipe that can trigger terminal commands

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] Add --control-pipe <path-to-pipe> option to `run --hot`
- [x] Strings written into the control pipe are passed to `handleTerminalCommand`

Launch:

```
flutter run --hot --control-pipe /tmp/my.pipe
```

Reload:

```
echo "r" > /tmp/my.pipe
```

Restart:

```
echo "R" > /tmp/my.pipe
```

### CODE REVIEW

**Comment by johnmccutchan** [comment]
@Hixie @HansMuller

File: packages/flutter_tools/lib/src/base/os.dart [hunk 1]
```diff
@@ -67,6 +70,12 @@ class _PosixUtils extends OperatingSystemUtils {
   void unzip(File file, Directory targetDirectory) {
     runSync(<String>['unzip', '-o', '-q', file.path, '-d', targetDirectory.path]);
   }
+
+  @override
+  File makePipe(String path) {
+    runSync(<String>['mkfifo', path]);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/os.dart]: please file a bug asking that we document our dependency on mkfifo

Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/base/os.dart]: https://github.com/flutter/flutter/issues/5284

**Comment by johnmccutchan** [comment]
PTAL, I've wired it up so all terminal commands can be sent to a control pipe

File: packages/flutter_tools/lib/src/hot.dart [hunk 2]
```diff
@@ -43,6 +44,24 @@ class HotRunner extends ResidentRunner {
   ApplicationPackage _package;
   String _mainPath;
   final AssetBundle bundle = new AssetBundle();
+  final File pipe;
+
+  Future<String> _readFromControlPipe() async {
+    Stream<List<int>> stream = pipe.openRead();
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/hot.dart]: nit: use 'final' on variables that you don't change again

File: packages/flutter_tools/lib/src/hot.dart [hunk 3]
```diff
@@ -43,6 +44,24 @@ class HotRunner extends ResidentRunner {
...
+    List<int> bytes = await stream.first;
+    return new String.fromCharCodes(bytes);
+  }
+
+  Future<Null> _startReadingFromControlPipe() async {
+    if (pipe == null) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/hot.dart]: nit: no braces on one line statements

File: packages/flutter_tools/lib/src/hot.dart [hunk 4]
```diff
@@ -43,6 +44,24 @@ class HotRunner extends ResidentRunner {
...
+      return;
+    }
+    while (true) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/hot.dart]: add a comment saying how this loop ends (does _readFromControlPipe eventually throw? or do we rely on calling "exit" to end programs that use this?)

File: packages/flutter_tools/lib/src/hot.dart [hunk 5]
```diff
@@ -176,16 +197,16 @@ class HotRunner extends ResidentRunner {
...
+  Future<Null> handleTerminalCommand(String code) async {
     final String lower = code.toLowerCase();
     if (lower == 'r' || code == AnsiTerminal.KEY_F5) {
       // F5, restart
       if (code == 'r') {
         // lower-case 'r'
-        _reloadSources();
+        await _reloadSources();
       } else {
         // upper-case 'R'.
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/hot.dart]: this code is wrong. it makes F5 into a cold restart. It should be a warm restart.

File: packages/flutter_tools/lib/src/hot.dart (hunk 2 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Done.

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 6]
```diff
@@ -104,20 +104,27 @@ abstract class ResidentRunner {
...
       return true;
     }
 
     return false;
   }
 
+  Future<Null> processTerminalInput(String command) async {
+    bool handled = await _commonTerminalInputHandler(command);
+    if (!handled) {
+      await handleTerminalCommand(command);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/resident_runner.dart]: nit: braces

File: packages/flutter_tools/lib/src/hot.dart (hunk 3 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Done.

File: packages/flutter_tools/lib/src/run.dart [hunk 7]
```diff
@@ -208,7 +208,7 @@ class RunAndStayResident extends ResidentRunner {
   }
 
   @override
-  void handleTerminalCommand(String code) {
+  Future<Null> handleTerminalCommand(String code) async {
     String lower = code.toLowerCase();
     if (lower == 'r' || code == AnsiTerminal.KEY_F5) {
       if (device.supportsRestart) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/run.dart]: why no await here where you call restart()?

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 8]
```diff
@@ -127,13 +135,29 @@ class RunCommand extends RunCommandBase {
         printError('Hot mode is not supported by this device.');
         return 1;
       }
+    } else {
+      if (argResults['control-pipe']) {
+        printError('--control-pipe requires --hot');
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/run.dart]: why? They seem like orthogonal features.

**Comment by Hixie** [comment]
LGTM with the nits and changes above.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 6 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Done.

File: packages/flutter_tools/lib/src/run.dart (hunk 7 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/run.dart]: Done.

File: packages/flutter_tools/lib/src/hot.dart (hunk 4 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Added a note about the special nature of this loop.

File: packages/flutter_tools/lib/src/hot.dart (hunk 5 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Fixed. 'r' and F5 now trigger _reloadSources. 'R' triggers _restartSources.

File: packages/flutter_tools/lib/src/commands/run.dart (hunk 8 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/commands/run.dart]: non-hot mode is going away.

Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/run.dart]: i thought we needed non-hot mode for --release and --profile modes?
